tests/integration/test_balancer.py

```diff
@@ -236,6 +236,95 @@ async def test_remove_unit_full_rebalance(self, ops_test: OpsTest, balancer_app)
+            idle_period=30,
+        )
+        async with ops_test.fast_forward(fast_interval="20s"):
+            await asyncio.sleep(120)  # ensure update-status adds broker-capacities if missed
+
+        assert balancer_is_ready(ops_test=ops_test, app_name=balancer_app)
+
+        await asyncio.sleep(30)  # let the API breathe after so many requests
+
+        # verify CC can find the new broker_id 3, with no replica partitions allocated
+        broker_replica_count = get_replica_count_by_broker_id(ops_test, balancer_app)
+        new_broker_id = max(map(int, broker_replica_count.keys()))
+        new_broker_replica_count = int(broker_replica_count.get(str(new_broker_id), 0))
+
+        assert not new_broker_replica_count
+
+        for unit in ops_test.model.applications[balancer_app].units:
+            if await unit.is_leader_from_status():
+                leader_unit = unit
+
+        rebalance_action_dry_run = await leader_unit.run_action(
+            "rebalance", mode="add", brokerid=[new_broker_id], dryrun=True, timeout=600, block=True
+        )
+        response = await rebalance_action_dry_run.wait()
+        assert response.results
+
+        rebalance_action = await leader_unit.run_action(
+            "rebalance",
+            mode="add",
+            brokerid=new_broker_id,
+            dryrun=False,
+            timeout=600,
+            block=True,
+        )
+        response = await rebalance_action.wait()
+        assert response.results
+
+        assert int(
+            get_replica_count_by_broker_id(ops_test, balancer_app).get(str(new_broker_id), 0)
+        )  # replicas were successfully moved
+
+    @pytest.mark.abort_on_fail
+    async def test_balancer_prepare_unit_removal(self, ops_test: OpsTest, balancer_app):
+        broker_replica_count = get_replica_count_by_broker_id(ops_test, balancer_app)
+        new_broker_id = max(map(int, broker_replica_count.keys()))
+
+        for unit in ops_test.model.applications[balancer_app].units:
+            if await unit.is_leader_from_status():
+                leader_unit = unit
+
+        rebalance_action_dry_run = await leader_unit.run_action(
+            "rebalance",
+            mode="remove",
+            brokerid=new_broker_id,
+            dryrun=True,
+            timeout=600,
+            block=True,
+        )
+        response = await rebalance_action_dry_run.wait()
+        assert response.results
+
+        rebalance_action = await leader_unit.run_action(
+            "rebalance",
+            mode="remove",
+            brokerid=[new_broker_id],
+            dryrun=False,
+            timeout=600,
+            block=True,
+        )
+        response = await rebalance_action.wait()
+        assert response.results
+
+        broker_replica_count = get_replica_count_by_broker_id(ops_test, balancer_app)
+        new_broker_replica_count = int(broker_replica_count.get(str(new_broker_id), 0))
+
+        assert not new_broker_replica_count
```

Comment:
**todo:** Similar to above comment, I would want to check that all leader/replicas are accounted for on the surviving units, and that they only go up or stay the same.

Reply:
Done in 23e85314453a

Not tested TBH, because it takes aaaaages